Add support for CogVLM model

Tianyue-Zhao opened the pull request:
This addresses the requests for CogVLM in #4387 and #4350.  
CogVLM is a pretty popular model that now adds in cleanly after the recent additions to libmtmd.  
I've converted a GGUF here: [Link to GGUF files](https://huggingface.co/PandaExpressPatron/cogvlm-chat-gguf/tree/main)  

Sample command and output:
```
build/bin/llama-mtmd-cli -m ../cogvlm-chat-hf/cogvlm-13B-chat-v1.1-F16.gguf --mmproj ../cogvlm-chat-hf/mmproj-cogvlm-chat-hf --image ./community.png --chat-template vicuna -p "Describe the picture"

load_hparams: model size:         8448.53 MiB
load_hparams: metadata size:      0.36 MiB
alloc_compute_meta:        CPU compute buffer size =   142.02 MiB
main: loading model: ../cogvlm-chat-hf/cogvlm-13B-chat-v1.1-F16.gguf
encoding image slice...
image slice encoded in 16135 ms
decoding image batch 1/1, n_tokens_batch = 1227
image decoded (batch 1/1) in 54065 ms

1. The image showcases a futuristic urban landscape with a mix of architectural styles. The buildings are multi-storied and have a combination of traditional and modern elements. There's a prominent tree in the foreground, suggesting a blend of nature and urban development. The scene appears to be bustling with activity, with various signs and billboards, indicating commercial or residential zones.


llama_perf_context_print:        load time =  108969.65 ms
llama_perf_context_print: prompt eval time =   85229.27 ms /  1241 tokens (   68.68 ms per token,    14.56 tokens per second)
llama_perf_context_print:        eval time =   19843.15 ms /    83 runs   (  239.07 ms per token,     4.18 tokens per second)
llama_perf_context_print:       total time =  126951.23 ms /  1324 tokens
llama_perf_context_print:    graphs reused =          0
```

- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md) 
- Self-reported review complexity:
  - [x] Low
  - [ ] Medium
  - [ ] High

Tianyue-Zhao committed:
Added GGUF mappings for CogVLM model

Tianyue-Zhao committed:
Add tensor mapping for CogVLM visual encoder

Tianyue-Zhao committed:
Add CogVLM to conversion script, no vision part yet

Tianyue-Zhao committed:
Added CogVLM vision model to conversion script

Tianyue-Zhao committed:
Add graph for CogVLM CLIP model

Tianyue-Zhao committed:
Add graph for CogVLM

Tianyue-Zhao committed:
Fixes for CogVLM. Now compiles.

Tianyue-Zhao committed:
Model now runs

Tianyue-Zhao committed:
Fixes for cogvlm graph

CISC commented:
> Also, is there a way to run these Github workflows locally or without needing approval from a reviewer? It would be good to run these CI/CD checks myself before posting the PR.

You can run `flake8`, `pyright` and `editorconfig` locally (or via IDE plugins), the build tests can be run manually with `ctest`.

CISC commented on convert_hf_to_gguf.py:
```diff
@@ -7928,6 +7928,62 @@ def prepare_tensors(self):
+
+@ModelBase.register("CogVLMForCausalLM")
+class CogVLMVisionModel(MmprojModel):
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.hparams['num_attention_heads'] = self.hparams['num_heads']
+
+    def set_gguf_parameters(self):
```
```suggestion
    def set_gguf_parameters(self):
```
Add `num_heads` to the list here instead:
https://github.com/ggml-org/llama.cpp/blob/de221570f9f3ed1bf352516525ea30b0fe73c5a5/convert_hf_to_gguf.py#L1258

Tianyue-Zhao replied:
So I didn't end up doing this, because after I add the "num_heads" to the list, self.find_vparam(["num_attention_heads"]) will still fail due to not finding the key.  
I think this workaround would need to stay... Unless there's some other way?

CISC replied:
In case it wasn't clear, I meant modifying the list like this: `self.find_vparam(["num_attention_heads", "num_heads"])`

Then it should just work and you don't need to change `hparams`.

CISC commented on convert_hf_to_gguf.py:
```diff
@@ -7928,6 +7928,62 @@ def prepare_tensors(self):
+        if not name.startswith("model.vision."):
+            return []
+
+        if "query_key_value" in name:
+            # Split tensor into three along first axis
+            q, k, v = data_torch.split(data_torch.shape[0] // 3, dim=0)
+            return [
+                (self.map_tensor_name(name.replace("query_key_value", "query")), q),
+                (self.map_tensor_name(name.replace("query_key_value", "key")), k),
+                (self.map_tensor_name(name.replace("query_key_value", "value")), v),
+            ]
+
+        return [(self.map_tensor_name(name), data_torch)]
```
```suggestion
        return [(self.map_tensor_name(name), data_torch)]
```
Create Q/K/V views at build time instead (check other (non-mm) models for examples).

CISC commented on convert_hf_to_gguf.py:
```diff
@@ -7928,6 +7928,62 @@ def prepare_tensors(self):
+class CogVLMModel(LlamaModel):
+    model_arch = gguf.MODEL_ARCH.COGVLM
+
+    def set_gguf_parameters(self):
+        super().set_gguf_parameters()
+
+    def modify_tensors(self, data_torch: Tensor, name: str, bid: int | None) -> Iterable[tuple[str, Tensor]]:
```
```suggestion
    def modify_tensors(self, data_torch: Tensor, name: str, bid: int | None) -> Iterable[tuple[str, Tensor]]:
```

Tianyue-Zhao replied:
Removed.

CISC commented on convert_hf_to_gguf.py:
```diff
@@ -7928,6 +7928,62 @@ def prepare_tensors(self):
+        if name.startswith("model.vision."):
+            return []
+
+        if "query_key_value.weight" in name:
+            # Slice tensor into three along first axis
+            q, k, v = data_torch.split(data_torch.shape[0] // 3, dim=0)
+            return [
+                (self.map_tensor_name(name.replace("query_key_value", "query")), q),
+                (self.map_tensor_name(name.replace("query_key_value", "key")), k),
+                (self.map_tensor_name(name.replace("query_key_value", "value")), v),
+            ]
+
+        return [(self.map_tensor_name(name), data_torch)]
```
```suggestion
        return [(self.map_tensor_name(name), data_torch)]
```

Tianyue-Zhao replied:
Same as above comment - both CLIP and the main model split the tensors in the graph instead.

CISC commented on src/llama-model.cpp:
```diff
@@ -17549,6 +17591,111 @@ struct llm_build_smallthinker : public llm_graph_context{
+                Kcur = ggml_reshape_3d(ctx0, Kcur, n_embd_head, n_head_kv, n_tokens);
+                Vcur = ggml_reshape_3d(ctx0, Vcur, n_embd_head, n_head_kv, n_tokens);
+
+                Qcur = ggml_rope(ctx0, Qcur, inp_pos, n_embd_head, GGML_ROPE_TYPE_NEOX);
+                Kcur = ggml_rope(ctx0, Kcur, inp_pos, n_embd_head, GGML_ROPE_TYPE_NEOX);
```
Update `llama_model_rope_type` instead and use `rope_type`.

CISC reviewed (commented):
This is not a complete review as I don't know enough about `mtmd`, just commenting...

Tianyue-Zhao commented:
> > Also, is there a way to run these Github workflows locally or without needing approval from a reviewer? It would be good to run these CI/CD checks myself before posting the PR.
> 
> You can run `flake8`, `pyright` and `editorconfig` locally (or via IDE plugins), the build tests can be run manually with `ctest`.

Thanks for the info! That's something I've been wondering about for a while.

CISC commented on tools/mtmd/clip.cpp:
```diff
@@ -1556,6 +1567,133 @@ struct clip_graph {
+        // Branch 2: multiply with mm_gate_w
+        ggml_tensor * gate = ggml_mul_mat(ctx0, model.mm_gate_w, cur);
+
+        // Apply silu
+        gate = ggml_silu_inplace(ctx0, gate);
+
+        // Multiply together
+        cur = ggml_mul(ctx0, gate, h_to_4h);
```
```suggestion
        // Apply swiglu
        cur = ggml_swiglu_split(ctx0, gate, h_to_4h);
```

Tianyue-Zhao replied:
Good point, I changed it to match how it's done in build_ffn.

CISC commented on src/llama-model.cpp:
```diff
@@ -18034,6 +18072,109 @@ struct llm_build_smallthinker : public llm_graph_context{
+                ggml_tensor * qkv = build_lora_mm(wqkv, cur);
+                cb(qkv, "qkv", il);
+
+                // split qkv into Q, K, V along the first dimension
+                ggml_tensor * Qcur = ggml_cont(ctx0, ggml_view_2d(ctx0, qkv, n_embd, n_tokens,
+                    qkv->nb[1], 0));
+                ggml_tensor * Kcur = ggml_cont(ctx0, ggml_view_2d(ctx0, qkv, n_embd, n_tokens,
+                    qkv->nb[1], n_embd * ggml_element_size(qkv)));
+                ggml_tensor * Vcur = ggml_cont(ctx0, ggml_view_2d(ctx0, qkv, n_embd, n_tokens,
+                    qkv->nb[1], 2 * n_embd * ggml_element_size(qkv)));
+
+                Qcur = ggml_reshape_3d(ctx0, Qcur, n_embd_head, n_head,    n_tokens);
+                Kcur = ggml_reshape_3d(ctx0, Kcur, n_embd_head, n_head_kv, n_tokens);
```
All backends can handle non-contiguous RoPE, so you don't need to `ggml_cont` Q/K here, and as a bonus you can directly create 3D views.

Tianyue-Zhao replied:
I changed it and indeed it works well, thanks for the comment.

Tianyue-Zhao committed:
Account for graph context change after rebase

Tianyue-Zhao committed:
Changes for whitespace

Tianyue-Zhao committed:
Changes in convert script according to comments

Tianyue-Zhao committed:
Switch CogVLM LLM graph to merged QKV tensor

Tianyue-Zhao committed:
Use rope_type variable instead of direct definition

Tianyue-Zhao committed:
Change CogVLM CLIP encoder to use SWIGLU

Tianyue-Zhao committed:
Switch CogVLM CLIP to use merged QKV

CISC reviewed (commented):
Further refinement (merge cont+reshape).

CISC commented:
> Further refinement (merge cont+reshape).

After #15662 we can avoid these altogether and just create 3D views.

CISC commented:
> > Further refinement (merge cont+reshape).
> 
> After #15662 we can avoid these altogether and just create 3D views.

Merged, rebase and apply updated suggestions.

Tianyue-Zhao committed:
Apply rebase edits and remove ggml_cont call that is now unnecessary

Tianyue-Zhao commented:
> > > Further refinement (merge cont+reshape).
> > 
> > 
> > After #15662 we can avoid these altogether and just create 3D views.
> 
> Merged, rebase and apply updated suggestions.

Thanks for the reminder, I've rebased it and removed the extra ggml_cont calls.

ngxson commented:
sorry I missed the notification to review this. will have a look & push commits to resolve the conflicts

ngxson committed:
Merge branch 'master' into cogvlm_support

ngxson committed:
clean up

ngxson reviewed (approved):
I don't have enough VRAM to test the model right now, but I think the code should be good to merge (after CI passed)

Feel free to give it a try even after the PR is merged. In case there are bugs, we can make follow-up PRs to fix it.
